photon-lib/src/main/scala/com/linkedin/photon/ml/function/PriorDistribution.scala

```diff
@@ -0,0 +1,285 @@
+ * under the License.
+ */
+package com.linkedin.photon.ml.function
+
+import breeze.linalg.{DenseMatrix, Vector, sum}
+import breeze.numerics.{abs, sqrt}
+
+import com.linkedin.photon.ml.constants.MathConst
+import com.linkedin.photon.ml.normalization.NormalizationContext
+import com.linkedin.photon.ml.util.BroadcastWrapper
+import com.linkedin.photon.ml.model.{Coefficients => ModelCoefficients}
+
+trait PriorDistribution extends ObjectiveFunction {
+
+  protected var _previousCoefficients: ModelCoefficients = _
+  protected var _l1RegWeight: Double = 0D
+  protected var _l2RegWeight: Double = 0D
+  // Todo: change prevHessian type to Matrix[Double] later
+  protected var prevMeans: Vector[Double] = _
+  protected var prevVariances: Vector[Double] = _
+  protected var prevHessian: Vector[Double] = _
+
+  /**
+   * Getter.
+   *
+   * @return The L1 regularization weight
+   */
+  def getL1RegularizationWeight: Double = _l1RegWeight
+
+  /**
+   * Getter.
+   *
+   * @return The L2 regularization weight
+   */
+  def getL2RegularizationWeight: Double = _l2RegWeight
+
+  /**
+   * Getter.
+   *
+   * @return The previous coefficients
+   */
+  def getPreviousCoefficients: ModelCoefficients = _previousCoefficients
+
+
+  /**
+   * Setter.
+   *
+   * @note This function definition uses the setter syntactic sugar trick. Statements like:
+   *
+   *    objectiveFunction.l1RegularizationWeight = 10
+   *
+   *       will call this function.
+   * @param newL1RegWeight The new L1 regularization weight
+   */
+  protected[ml] def l1RegularizationWeight_=(newL1RegWeight: Double): Unit = _l1RegWeight = newL1RegWeight
+
+  /**
+   * Setter.
+   *
+   * @note This function definition uses the setter syntactic sugar trick. Statements like:
+   *
+   *    objectiveFunction.l2RegularizationWeight = 10
+   *
+   *       will call this function.
+   * @param newL2RegWeight The new L2 regularization weight
+   */
+  protected[ml] def l2RegularizationWeight_=(newL2RegWeight: Double): Unit = _l2RegWeight = newL2RegWeight
+
+  /**
+   * Setter.
+   *
+   * @note This function definition uses the setter syntactic sugar trick. Statements like:
+   *
+   *    objectiveFunction.previousCoefficients_ = newPreviousCoefficients
+   *
+   *       will call this function.
+   * @param newPreviousCoefficients The new previous coefficients
+   */
+  protected[ml] def previousCoefficients_=(newPreviousCoefficients: ModelCoefficients): Unit  = {
+
+    _previousCoefficients = newPreviousCoefficients
+    val priorDistributionStats = _previousCoefficients match {
+      case ModelCoefficients(means: Vector[Double], Some(variances: Vector[Double])) => (means, variances)
+      case ModelCoefficients(means: Vector[Double], None) =>
+        throw new IllegalArgumentException("Previous variances should not be empty.")
+    }
+    prevMeans = priorDistributionStats._1
+    prevVariances = priorDistributionStats._2
```

Comment:
You can avoid need lines 99 and 100 by changing line 94 to:
```
(prevMeans, prevVariances) = _previousCoefficients match { ... }
```

Reply:
This will cause error ```PriorDistribution.scala:98: ';' expected but '=' found.
    (_prevMeans, _prevVariances) = _previousCoefficients match {
```